lib/tool_shed/util/shed_util_common.py

```diff
@@ -604,36 +604,36 @@ def get_repository_for_dependency_relationship( app, tool_shed, name, owner, cha
-    if not repository:
-        tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
-        repository_clone_url = os.path.join( tool_shed_url, 'repos', owner, name )
-        repo_info_tuple = (None, repository_clone_url, changeset_revision, None, owner, None, None)
-        repository, pcr = repository_was_previously_installed( app, tool_shed_url, name, repo_info_tuple )
-    if not repository:
-        # The received changeset_revision is no longer installable, so get the next changeset_revision
-        # in the repository's changelog in the tool shed that is associated with repository_metadata.
-        tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
-        params = dict( name=name, owner=owner, changeset_revision=changeset_revision )
-        pathspec = [ 'repository', 'next_installable_changeset_revision' ]
-        text = util.url_get( tool_shed_url, password_mgr=app.tool_shed_registry.url_auth( tool_shed_url ), pathspec=pathspec, params=params )
-        if text:
-            repository = get_installed_repository( app=app,
-                                                   tool_shed=tool_shed,
-                                                   name=name,
-                                                   owner=owner,
-                                                   changeset_revision=text )
-    return repository
+    repositories = get_installed_repository( app=app,
+                                             tool_shed=tool_shed,
+                                             name=name,
+                                             owner=owner,
+                                             return_all=True)
+    if repositories:
+        if repositories[0].installed_changeset_revision == changeset_revision:
+            repository = repositories[0]
+        else:
+            changeset_repo_d = { repo.changeset_revision: repo for repo in repositories }
+            repository = changeset_repo_d.get( changeset_revision, None )
+        if not repository:
+            tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
+            changeset_revisions = get_previous_changeset_revisions(app=app,
+                                                                   tool_shed_url=tool_shed_url,
+                                                                   name=name,
+                                                                   owner=owner,
+                                                                   changeset_revision=changeset_revision)
+            for previous_changeset in changeset_revisions:
+                if previous_changeset in changeset_repo_d:
+                    return changeset_repo_d[ previous_changeset ]
+            # The received changeset_revision is no longer installable, so get the next changeset_revision
```

Comment:
@mvdbeek Thanks for taking a look at this.  Its been a while since I was involved with this particular code, so please take my input with a grain of salt or two.  ;)

I assume the scenario that results in this function receiving an older changeset_revision than the one that is currently installed is that a repository is being installed that has a dependency on the older changeset_revision.  If this is the case, the older changeset_revision should be installed unless the repository type is tool_dependency_definition.  Since this function is called for all repository types, I think the repository type should be checked so that your new approach is used only for repositories of type tool_dependency_definition.  If the type is unrestricted, I think the original approach should be used because that scenario requires installation of the older changeset_revision.

I like the approach you've taken with the optional param for returning all installed repositories, no matter the changeset_revision.